quarkus/runtime/src/main/java/org/keycloak/QuarkusKeycloakApplication.java

```diff
@@ -1,10 +1,402 @@
+import org.jboss.logging.Logger;
+import org.keycloak.common.util.Resteasy;
+import org.keycloak.config.ConfigProviderFactory;
+import org.keycloak.exportimport.ExportImportManager;
+import org.keycloak.migration.MigrationModelManager;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.KeycloakSessionTask;
+import org.keycloak.models.ModelDuplicateException;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.UserProvider;
+import org.keycloak.models.dblock.DBLockManager;
+import org.keycloak.models.dblock.DBLockProvider;
+import org.keycloak.models.utils.KeycloakModelUtils;
+import org.keycloak.models.utils.PostMigrationEvent;
+import org.keycloak.models.utils.RepresentationToModel;
+import org.keycloak.platform.Platform;
+import org.keycloak.platform.PlatformProvider;
+import org.keycloak.representations.idm.RealmRepresentation;
+import org.keycloak.representations.idm.UserRepresentation;
+import org.keycloak.services.DefaultKeycloakSessionFactory;
+import org.keycloak.services.ServicesLogger;
+import org.keycloak.services.error.KeycloakErrorHandler;
+import org.keycloak.services.filters.KeycloakSecurityHeadersFilter;
+import org.keycloak.services.filters.KeycloakTransactionCommitter;
+import org.keycloak.services.managers.ApplianceBootstrap;
+import org.keycloak.services.managers.RealmManager;
+import org.keycloak.services.managers.UserStorageSyncManager;
+import org.keycloak.services.resources.JsResource;
 import org.keycloak.services.resources.KeycloakApplication;
+import org.keycloak.services.resources.RealmsResource;
+import org.keycloak.services.resources.RobotsResource;
+import org.keycloak.services.resources.ThemeResource;
+import org.keycloak.services.resources.WelcomeResource;
+import org.keycloak.services.resources.admin.AdminRoot;
+import org.keycloak.services.scheduled.ClearExpiredClientInitialAccessTokens;
+import org.keycloak.services.scheduled.ClearExpiredEvents;
+import org.keycloak.services.scheduled.ClearExpiredUserSessions;
+import org.keycloak.services.scheduled.ClusterAwareScheduledTaskRunner;
+import org.keycloak.services.scheduled.ScheduledTaskRunner;
+import org.keycloak.services.util.ObjectMapperResolver;
+import org.keycloak.timer.TimerProvider;
+import org.keycloak.transaction.JtaTransactionManagerLookup;
+import org.keycloak.util.JsonSerialization;
 
 @ApplicationPath("/")
 public class QuarkusKeycloakApplication extends KeycloakApplication {
 
+    private static final Logger logger = Logger.getLogger(KeycloakApplication.class);
+
+    protected final PlatformProvider platform = Platform.getPlatform();
+
+    protected Set<Object> singletons = new HashSet<Object>();
+    protected Set<Class<?>> classes = new HashSet<Class<?>>();
+
+    protected KeycloakSessionFactory sessionFactory;
+
+    public QuarkusKeycloakApplication() {
+
+        try {
```

Comment:
Would it be possible to move the initialization logic into a dedicated protected init() method?

Reply:
It is a simple change. But that is subject to change. Short term, we are planning quite a few other changes that might impact your code. E.g.; master realm initialization shall move to another place. We should really avoid spreading bootstrap code across different places.

But I have no objection :) @stianst ?

Reply:
`Short term, we are planning quite a few other changes that might impact your code.`
I'm aware of that :) Sometimes, it's hard to keep up with your speed in my spare-time :D

I'm asking, because the initialization within this constructor requires a bunch of infrastructure to be in place already: The required components need to be either statically configured or populated in a JNDI environment.
It would be much nicer if the dependencies needed by the KeycloakApplication instance could be passed in as constructor arguments directly, or indirectly via some kind of Context object that's passed in. 

This would allow for more control about when which required component is created. 
Another benefit would be to ease subclassing the `KeycloakApplication` with more control about the component bootstrap there. 

IIRC it is possible to configure the `org.jboss.resteasy.plugins.server.servlet.HttpServlet30Dispatcher` with a predefined `javax.ws.rs.Application` instance instead of just providing it with a FQCN. This could also ease testing of KeycloakApplication components due to easier mocking.

The `HttpServlet30Dispatcher` could configured, deployed programmatically via a `ServletContextListener` which would then bootstrap the whole environment.

Reply:
The thing is that we are dropping Servlet API for Quarkus. We are now using Vert.x.

We are also looking forward to making `KeycloakApplication` really specific to bootstrapping JAX-RS whereas the session factory is basically a singleton that can be initialized from itself.

I guess you are doing this for your Keycloak On SB project, right? If we just make the session factory doing all the heavy work you can easily bootstrap the server, from everywhere.

Reply:
In this PR we're just making a copy of the KeycloakApplication in order to make more drastic changes here without affecting the main distribution.

Let's discuss this on the mailing list instead of this PR.